File: explicit_one_step/driver.cpp
/* Main routine to test the higher-order one-step methods

   D.R. Reynolds
   Math 6321 @ SMU
   Fall 2020 */

#include <iostream>
#include "heun.hpp"
#include "erk2.hpp"
#include "erk3.hpp"
#include "erk4.hpp"
#include "taylor2.hpp"

using namespace std;
using namespace arma;

class RHS: public RHSFunction {
public:
  int Evaluate(double t, vec& y, vec& f) {    // evaluates the RHS function, f(t,y)
    f = -y*exp(-t);
    return 0;
  }
};

class RHSt: public RHSFunction {
public:
  int Evaluate(double t, vec& y, vec& ft) {    // evaluates the RHS function, f_t(t,y)
    ft = y*exp(-t);
    return 0;
  }
};

class RHSy: public RHSJacobian {
public:
  int Evaluate(double t, vec& y, mat& fy) {    // evaluates the RHS Jacobian, f_y(t,y)
    fy(0) = -exp(-t);
    return 0;
  }
};

// Convenience function for analytical solution
vec ytrue(const double t) {
  vec yt(1);
  yt(0) = exp(exp(-t)-1.0);
  return yt;
};


// main routine
int main() {

  // time steps to try
  vec h("0.5 0.1 0.05 0.01 0.005 0.001 0.0005");

  // vector for errors at each h
  vec e(h.n_elem);

  // set problem information
  vec y0("1.0");
  double t0 = 0.0;
  double Tf = 1.0;

  // set desired output times
  int Nout = 3;  // includes initial condition
  vec tspan = linspace(t0, Tf, Nout);

  // create ODE RHS function objects
  RHS  f;
  RHSt ft;
  RHSy fy;

  // create true solution results
  mat Ytrue(1,Nout);
  for (size_t i=0; i<Nout; i++)
    Ytrue.col(i) = ytrue(tspan(i));
  
  //---- Taylor 2 ----
  cout << "\nTaylor2:\n";
  Taylor2Stepper T2(f, ft, fy, y0);
  for (size_t ih=0; ih<h.n_elem; ih++) {

    // call stepper
    cout << "  h = " << h(ih) << ":";
    mat Y = T2.Evolve(tspan, h(ih), y0);

    // output solution, errors, and overall error
    mat Yerr = abs(Y-Ytrue);
    e(ih) = Yerr.max();
    if (ih > 0) {
      cout << "  Max error = " << e(ih) << ",  conv rate = " 
           << log(e(ih)/e(ih-1))/log(h(ih)/h(ih-1)) << endl;
    } else {
      cout << "  Max error = " << e(ih) << endl;
    }

  }


  //---- Heun ----
  cout << "\nHeun:\n";
  HeunStepper H(f, y0);
  for (size_t ih=0; ih<h.n_elem; ih++) {

    // call stepper
    cout << "  h = " << h(ih) << ":";
    mat Y = H.Evolve(tspan, h(ih), y0);

    // output solution, errors, and overall error
    mat Yerr = abs(Y-Ytrue);
    e(ih) = Yerr.max();
    if (ih > 0) {
      cout << "  Max error = " << e(ih) << ",  conv rate = " 
           << log(e(ih)/e(ih-1))/log(h(ih)/h(ih-1)) << endl;
    } else {
      cout << "  Max error = " << e(ih) << endl;
    }

  }


  //---- ERK 2 ----
  cout << "\nERK2:\n";
  ERK2Stepper ERK2(f, y0);
  for (size_t ih=0; ih<h.n_elem; ih++) {

    // call stepper
    cout << "  h = " << h(ih) << ":";
    mat Y = ERK2.Evolve(tspan, h(ih), y0);

    // output solution, errors, and overall error
    mat Yerr = abs(Y-Ytrue);
    e(ih) = Yerr.max();
    if (ih > 0) {
      cout << "  Max error = " << e(ih) << ",  conv rate = " 
           << log(e(ih)/e(ih-1))/log(h(ih)/h(ih-1)) << endl;
    } else {
      cout << "  Max error = " << e(ih) << endl;
    }

  }


  //---- ERK3 ----
  cout << "\nERK3:\n";
  ERK3Stepper ERK3(f, y0);
  for (size_t ih=0; ih<h.n_elem; ih++) {

    // call stepper
    cout << "  h = " << h(ih) << ":";
    mat Y = ERK3.Evolve(tspan, h(ih), y0);

    // output solution, errors, and overall error
    mat Yerr = abs(Y-Ytrue);
    e(ih) = Yerr.max();
    if (ih > 0) {
      cout << "  Max error = " << e(ih) << ",  conv rate = " 
           << log(e(ih)/e(ih-1))/log(h(ih)/h(ih-1)) << endl;
    } else {
      cout << "  Max error = " << e(ih) << endl;
    }

  }


  //---- ERK4 ----
  cout << "\nERK4:\n";
  ERK4Stepper ERK4(f, y0);
  for (size_t ih=0; ih<h.n_elem; ih++) {

    // call stepper
    cout << "  h = " << h(ih) << ":";
    mat Y = ERK4.Evolve(tspan, h(ih), y0);

    // output solution, errors, and overall error
    mat Yerr = abs(Y-Ytrue);
    e(ih) = Yerr.max();
    if (ih > 0) {
      cout << "  Max error = " << e(ih) << ",  conv rate = " 
           << log(e(ih)/e(ih-1))/log(h(ih)/h(ih-1)) << endl;
    } else {
      cout << "  Max error = " << e(ih) << endl;
    }

  }


  return 0;
}

File: explicit_one_step/driver.py
#!/usr/bin/env python3
#
# Main routine to test the higher-order one-step methods.
#
# D.R. Reynolds
# Math 6321 @ SMU
# Fall 2023

import numpy as np
from Taylor2 import *
from ERK import *

# problem time interval
t0 = 0.0
tf = 1.0

# problem-definining functions and initial conditions
def f(t,y):
    """ ODE RHS function """
    return -y*np.exp(-t)
def f_t(t,y):
    """ t-derivative of ODE RHS function """
    return y*np.exp(-t)
def f_y(t,y):
    """ y-derivative of ODE RHS function """
    return np.array([[-np.exp(-t)]])
def ytrue(t):
    """ Analytical solution """
    return np.exp(np.exp(-t)-1.0)

# shared testing data
Nout = 3   # includes initial condition
tspan = np.linspace(t0, tf, Nout)

# create true solution results
Ytrue = np.zeros((Nout,1))
for i in range(Nout):
    Ytrue[i,:] = ytrue(tspan[i])

# time steps to try
hvals = np.array([0.5, 0.1, 0.05, 0.01, 0.005, 0.001, 0.0005])
errs = np.zeros(hvals.size)

#### Taylor 2 ####
print("\nTaylor2:")
T2 = Taylor2(f, f_t, f_y)
for idx, h in enumerate(hvals):

    # set initial condition and call stepper
    y0 = Ytrue[0,:]
    print("  h = ", h, ":")
    T2.reset()
    Y, success = T2.Evolve(tspan, y0, h)

    # output solution, errors, and overall error
    Yerr = np.abs(Y-Ytrue)
    errs[idx] = np.linalg.norm(Yerr,np.inf)
    for i in range(Nout):
        print("    y(%.1f) = %9.6f   |error| = %.2e" % (tspan[i], Y[i,0], Yerr[i,0]))
    print("  overall:  steps = %5i  nrhs = %5i  abserr = %9.2e  relerr = %9.2e\n" %
          (T2.get_num_steps(), T2.get_num_rhs(), errs[idx], np.linalg.norm(Yerr/Ytrue,np.inf)))
orders = np.log(errs[0:-2]/errs[1:-1])/np.log(hvals[0:-2]/hvals[1:-1])
print('estimated order: max = ', np.max(orders), ',  avg = ', np.average(orders))

#### Heun ####
print("\nHeun:")
A, b, c, p = Heun()
H = ERK(f, A, b, c)
for idx, h in enumerate(hvals):

    # set initial condition and call stepper
    y0 = Ytrue[0,:]
    print("  h = ", h, ":")
    H.reset()
    Y, success = H.Evolve(tspan, y0, h)

    # output solution, errors, and overall error
    Yerr = np.abs(Y-Ytrue)
    errs[idx] = np.linalg.norm(Yerr,np.inf)
    for i in range(Nout):
        print("    y(%.1f) = %9.6f   |error| = %.2e" % (tspan[i], Y[i,0], Yerr[i,0]))
    print("  overall:  steps = %5i  nrhs = %5i  abserr = %9.2e  relerr = %9.2e\n" %
          (H.get_num_steps(), H.get_num_rhs(), errs[idx], np.linalg.norm(Yerr/Ytrue,np.inf)))
orders = np.log(errs[0:-2]/errs[1:-1])/np.log(hvals[0:-2]/hvals[1:-1])
print('estimated order: max = ', np.max(orders), ',  avg = ', np.average(orders))

#### ERK2 ####
print("\nERK2:")
A, b, c, p = ERK2()
E2 = ERK(f, A, b, c)
for idx, h in enumerate(hvals):

    # set initial condition and call stepper
    y0 = Ytrue[0,:]
    print("  h = ", h, ":")
    E2.reset()
    Y, success = E2.Evolve(tspan, y0, h)

    # output solution, errors, and overall error
    Yerr = np.abs(Y-Ytrue)
    errs[idx] = np.linalg.norm(Yerr,np.inf)
    for i in range(Nout):
        print("    y(%.1f) = %9.6f   |error| = %.2e" % (tspan[i], Y[i,0], Yerr[i,0]))
    print("  overall:  steps = %5i  nrhs = %5i  abserr = %9.2e  relerr = %9.2e\n" %
          (E2.get_num_steps(), E2.get_num_rhs(), errs[idx], np.linalg.norm(Yerr/Ytrue,np.inf)))
orders = np.log(errs[0:-2]/errs[1:-1])/np.log(hvals[0:-2]/hvals[1:-1])
print('estimated order: max = ', np.max(orders), ',  avg = ', np.average(orders))

#### ERK3 ####
print("\nERK3:")
A, b, c, p = ERK3()
E3 = ERK(f, A, b, c)
for idx, h in enumerate(hvals):

    # set initial condition and call stepper
    y0 = Ytrue[0,:]
    print("  h = ", h, ":")
    E3.reset()
    Y, success = E3.Evolve(tspan, y0, h)

    # output solution, errors, and overall error
    Yerr = np.abs(Y-Ytrue)
    errs[idx] = np.linalg.norm(Yerr,np.inf)
    for i in range(Nout):
        print("    y(%.1f) = %9.6f   |error| = %.2e" % (tspan[i], Y[i,0], Yerr[i,0]))
    print("  overall:  steps = %5i  nrhs = %5i  abserr = %9.2e  relerr = %9.2e\n" %
          (E3.get_num_steps(), E3.get_num_rhs(), errs[idx], np.linalg.norm(Yerr/Ytrue,np.inf)))
orders = np.log(errs[0:-2]/errs[1:-1])/np.log(hvals[0:-2]/hvals[1:-1])
print('estimated order: max = ', np.max(orders), ',  avg = ', np.average(orders))

#### ERK4 ####
print("\nERK4:")
A, b, c, p = ERK4()
E4 = ERK(f, A, b, c)
for idx, h in enumerate(hvals):

    # set initial condition and call stepper
    y0 = Ytrue[0,:]
    print("  h = ", h, ":")
    E4.reset()
    Y, success = E4.Evolve(tspan, y0, h)

    # output solution, errors, and overall error
    Yerr = np.abs(Y-Ytrue)
    errs[idx] = np.linalg.norm(Yerr,np.inf)
    for i in range(Nout):
        print("    y(%.1f) = %9.6f   |error| = %.2e" % (tspan[i], Y[i,0], Yerr[i,0]))
    print("  overall:  steps = %5i  nrhs = %5i  abserr = %9.2e  relerr = %9.2e\n" %
          (E4.get_num_steps(), E4.get_num_rhs(), errs[idx], np.linalg.norm(Yerr/Ytrue,np.inf)))
orders = np.log(errs[0:-2]/errs[1:-1])/np.log(hvals[0:-2]/hvals[1:-1])
print('estimated order: max = ', np.max(orders), ',  avg = ', np.average(orders))
